In [143]:
#import packages
import pandas as pd
import numpy as np
import geopandas as gpd
import matplotlib.pyplot as plt 


In [144]:
# pandas version 1.1.2 works with  large data sets.
pd.__version__

'1.1.2'

In [145]:
import os
directory=os.getcwd()
print(directory)


/Users/kultiginbozdemir/Documents/GitHub/vytal/angelo


In [146]:
customer_postgre_df=pd.read_csv('Customer Info Data PostGre.csv', sep=',', index_col='UserId')
customer_postgre_df.head()

LastActionTimestamp  \
UserId                                                                
b68ceac8-fa36-4149-a302-d1ce25154172  2021-10-24 16:30:28.393575+00   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613         0001-01-01 00:00:00+00   
9eec3024-2a26-4b04-8f84-74ab5f301028         0001-01-01 00:00:00+00   
5c630b06-2e80-491f-ad09-ed130936ff47  2020-11-18 07:05:21.876689+00   
4a730e8d-dd78-444a-8da0-ac8256f10f21  2021-10-20 15:45:35.417341+00   

                                                            AppSessionRefreshToken  \
UserId                                                                               
b68ceac8-fa36-4149-a302-d1ce25154172  /UlOcx0cPEJNuuwyoga8FbvjyE7uHqhAwaGgRC/wEhQ=   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613  UXqt3xl8bJhFT3jmb4Iv7O1yriaOS90dckDdad757zM=   
9eec3024-2a26-4b04-8f84-74ab5f301028  w/WIpnSFFJc1/5gOVMCzV8YNH9BHiHjbogNEO71BTC0=   
5c630b06-2e80-491f-ad09-ed130936ff47  awHAVO6nGdP3lH4UF1/lVXmNnZSU+M/plm+YaNMpKe4=   
4a730e8d-dd78-444a-8da0-ac8256f10f21  UM6bHyW+D8yaPwKIpF48Z+Yfmx/eXiu7UrMr8I2OQaI=   

                                               RefreshTokenIssuedAt  \
UserId                                                                
b68ceac8-fa36-4149-a302-d1ce25154172  2020-10-28 17:12:52.514232+00   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613  2020-10-28 17:13:40.177048+00   
9eec3024-2a26-4b04-8f84-74ab5f301028  2021-06-24 09:32:37.168007+00   
5c630b06-2e80-491f-ad09-ed130936ff47  2020-08-30 21:11:39.561842+00   
4a730e8d-dd78-444a-8da0-ac8256f10f21  2021-08-07 15:00:07.523963+00   

                                     SignupCodeInfo_SignupCode  \
UserId                                                           
b68ceac8-fa36-4149-a302-d1ce25154172                       NaN   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613                       NaN   
9eec3024-2a26-4b04-8f84-74ab5f301028                       NaN   
5c630b06-2e80-491f-ad09-ed130936ff47                       NaN   
4a730e8d-dd78-444a-8da0-ac8256f10f21                       NaN   

                                      SignupCodeInfo_SignupCodeWasUsed  \
UserId                                                                   
b68ceac8-fa36-4149-a302-d1ce25154172                             False   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613                             False   
9eec3024-2a26-4b04-8f84-74ab5f301028                             False   
5c630b06-2e80-491f-ad09-ed130936ff47                             False   
4a730e8d-dd78-444a-8da0-ac8256f10f21                             False   

                                      IsConvertedMobileUser  \
UserId                                                        
b68ceac8-fa36-4149-a302-d1ce25154172                  False   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613                  False   
9eec3024-2a26-4b04-8f84-74ab5f301028                  False   
5c630b06-2e80-491f-ad09-ed130936ff47                  False   
4a730e8d-dd78-444a-8da0-ac8256f10f21                  False   

                                      IsConvertedOfflineUser  IsMobileUser  \
UserId                                                                       
b68ceac8-fa36-4149-a302-d1ce25154172                   False         False   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613                   False         False   
9eec3024-2a26-4b04-8f84-74ab5f301028                   False         False   
5c630b06-2e80-491f-ad09-ed130936ff47                   False         False   
4a730e8d-dd78-444a-8da0-ac8256f10f21                   False         False   

                                      IsOfflineUser  Referrer  ...  \
UserId                                                         ...   
b68ceac8-fa36-4149-a302-d1ce25154172          False       NaN  ...   
b2e0a6ca-2f71-4d1e-8f66-48020f2db613          False       NaN  ...   
9eec3024-2a26-4b04-8f84-74ab5f301028          False       NaN  ...   
5c630b06-2e80-491f-ad09-ed130936ff47          False       NaN  ...   
4a730e8d-dd78-444a-8da0-ac8256f10f21          False       

In [147]:
# Import customer list
customer_2021_df=pd.read_csv('customers-2021-11-16_13-21.csv', sep=',', index_col='id')
customer_2021_df.head()

/Users/kultiginbozdemir/opt/anaconda3/envs/CO2/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3147: DtypeWarning: Columns (1,14,20,24,30,38,40,43,44,47) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Partner accept_email_functional  \
id                                                                     
f4977b0c-c05e-4098-aa54-07eecd160b4e     NaN                     NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2     NaN                     NaN   
d14fe6ba-fa96-432e-885e-6896db943e82     NaN                     NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e     NaN                     NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d     NaN                     NaN   

                                     accept_email_marketing  \
id                                                            
f4977b0c-c05e-4098-aa54-07eecd160b4e                    NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2                    NaN   
d14fe6ba-fa96-432e-885e-6896db943e82                    NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e                    NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                    NaN   

                                     accept_push_functional  \
id                                                            
f4977b0c-c05e-4098-aa54-07eecd160b4e                    NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2                    NaN   
d14fe6ba-fa96-432e-885e-6896db943e82                    NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e                    NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                    NaN   

                                     accept_push_marketing account_type  \
id                                                                        
f4977b0c-c05e-4098-aa54-07eecd160b4e                   NaN          NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2                   NaN          NaN   
d14fe6ba-fa96-432e-885e-6896db943e82                   NaN          NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e                   NaN          NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                   NaN          NaN   

                                      accounts_for_mail_address  \
id                                                                
f4977b0c-c05e-4098-aa54-07eecd160b4e                        2.0   
556b68e7-e872-4720-9efc-aeaf7159d6a2                        2.0   
d14fe6ba-fa96-432e-885e-6896db943e82                        2.0   
c4bade35-f576-4acd-9ba2-6705d60e036e                        1.0   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                        1.0   

                                      activated_campaign_count  \
id                                                               
f4977b0c-c05e-4098-aa54-07eecd160b4e                       NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2                       NaN   
d14fe6ba-fa96-432e-885e-6896db943e82                       NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e                       NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                       NaN   

                                     activated_campaigns activation_journey  \
id                                                                            
f4977b0c-c05e-4098-aa54-07eecd160b4e                 NaN                NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2                 NaN                NaN   
d14fe6ba-fa96-432e-885e-6896db943e82                 NaN                NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e                 NaN                NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d                 NaN                NaN   

                                      ...  promo_order_count  proposed_store  \
id                                    ...                                      
f4977b0c-c05e-4098-aa54-07eecd160b4e  ...                NaN             NaN   
556b68e7-e872-4720-9efc-aeaf7159d6a2  ...                NaN             NaN   
d14fe6ba-fa96-432e-885e-6896db943e82  ...                NaN             NaN   
c4bade35-f576-4acd-9ba2-6705d60e036e  ...                NaN             NaN   
e27c9d35-24af-4b2c-b2a7-3ac886cc225d  ...                NaN             NaN   

                                      proposed_store_name push_subscriptio

In [148]:
# Import container list
con_df=pd.read_csv('../data/Containerlist.csv', sep=',', index_col='Id')
con_df.head()

Name  TypeId  \
Id                                                             
00004f6b-e416-4a86-91df-5d177520d3a0  Marinha Grande       1   
00005175-379b-458a-9740-7741814674e4          Dayton       1   
0000bbe4-2051-4c32-850f-220288b868a6           Nahal       6   
0000c7c3-8f3c-4eb2-b586-dcae82555d16      Winfred F.       1   
0000f63c-9747-4c15-871f-f011198c8f88    Billy Wilder       3   

                                                               CreatedById  \
Id                                                                           
00004f6b-e416-4a86-91df-5d177520d3a0  cb73faf3-410a-4e71-83cb-c303efeaeba8   
00005175-379b-458a-9740-7741814674e4  fd3ec806-b763-48a9-992b-09e50c8a274c   
0000bbe4-2051-4c32-850f-220288b868a6                                   NaN   
0000c7c3-8f3c-4eb2-b586-dcae82555d16  cb73faf3-410a-4e71-83cb-c303efeaeba8   
0000f63c-9747-4c15-871f-f011198c8f88  fd3ec806-b763-48a9-992b-09e50c8a274c   

                                     ShortId  
Id                                            
00004f6b-e416-4a86-91df-5d177520d3a0  61CQAR  
00005175-379b-458a-9740-7741814674e4     NaN  
0000bbe4-2051-4c32-850f-220288b868a6     NaN  
0000c7c3-8f3c-4eb2-b586-dcae82555d16  Z9HBTP  
0000f63c-9747-4c15-871f-f011198c8f88     NaN

In [149]:
# Import container typs
c_type_df=pd.read_csv('../data/ContainerTypes.csv', sep=',', index_col='Id')
print(c_type_df.shape)
c_type_df.head()

(16, 4)


Name                Description  CategoryId  IsWhiteLabel
Id                                                                  
1   1250ml Bowl            A standard bowl         1.0         False
2    500ml Bowl             A smaller Bowl         1.0         False
3    750ml Bowl              A medium bowl         1.0         False
4      Sushibox            A box for sushi         1.0         False
5           Cup  A reusable cocktail glass         1.0         False

In [150]:
# Import merchants
m_df=pd.read_csv('Merchant_info.csv', sep=',', index_col='Id')
m_df['CreationDate']= pd.to_datetime(m_df['CreationDate'])
m_df.head()


MerchantId  \
Id                                                                           
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2  f0da3333-e946-467a-9230-6a6a913592b3   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5  901c6321-f762-46ed-8d47-677a3002f3c0   
00134510-51ae-45ef-98b6-05a6914c5f09  f2a1c384-cbb6-45c5-adf2-70eff8dc167a   
001c9d74-234f-4847-8411-2bb5b0c6480c  7fb259d7-c730-46b0-bdbf-a00f1df674a9   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e  4c62d324-a7ed-4fac-9cf5-dbd9e688dda7   

                                      IsRemoved  \
Id                                                
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2      False   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5      False   
00134510-51ae-45ef-98b6-05a6914c5f09      False   
001c9d74-234f-4847-8411-2bb5b0c6480c      False   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e      False   

                                                                                  Name  \
Id                                                                                       
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2                      Habibi und Hawara Landstraße   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5  Cafeteria im Behördenzentrum (Finanzministerium)   
00134510-51ae-45ef-98b6-05a6914c5f09                                          Nem Viet   
001c9d74-234f-4847-8411-2bb5b0c6480c                   Seligweiler Hotel & Restaurants   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e                             Fuchswinkel Vohwinkel   

                                      Longitude   Latitude  \
Id                                                           
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2  16.393554  48.202837   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5  11.058662  50.967486   
00134510-51ae-45ef-98b6-05a6914c5f09   6.956255  50.933111   
001c9d74-234f-4847-8411-2bb5b0c6480c  10.031569  48.457239   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e   7.068608  51.231759   

                                                   CreationDate  CategoryId  \
Id                                                                            
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2 2021-07-18 18:01:41.234055           2   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5 2021-08-02 07:04:12.313362           4   
00134510-51ae-45ef-98b6-05a6914c5f09 2021-04-28 12:14:49.087604           2   
001c9d74-234f-4847-8411-2bb5b0c6480c 2021-05-28 09:11:03.750535           2   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e 2021-07-20 11:54:36.199851           5   

                                                                         FoodInfoTextDe  \
Id                                                                                        
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2  Geschmack kennt keine Grenzen: Habibi & Hawara...   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5  Montag bis Freitag schnelles und kompliziertes...   
00134510-51ae-45ef-98b6-05a6914c5f09                                                NaN   
001c9d74-234f-4847-8411-2bb5b0c6480c  Unser Jakobs Grill ist eines der beliebtesten ...   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e                                                NaN   

                                                                         FoodInfoTextEn  \
Id                                                                                        
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2                                                NaN   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5                                                NaN   
00134510-51ae-45ef-98b6-05a6914c5f09                                                NaN   
001c9d74-234f-4847-8411-2bb5b0c6480c  Our Jakobs Grill is one of the most popular st...   
002deecc-8e00-4902-b5f3-d4fbd26a9b1e                                                NaN   

                                          PhoneNumber  ...  \
Id                                                     ...   
000b52b3-5a1b-4ce3-9895-ec67dcb80cd2       01/7108727  ...   
0011d510-7b42-4a99-b0cb-6be4d0cf23a5   +49361 2629197  ...   
00134510-51ae-4

In [151]:
min(m_df['CreationDate'])

Timestamp('2019-09-08 10:22:48.441084')

In [152]:
m_df.groupby(by='MerchantId').size().sort_values()

MerchantId
7e286913-c2b3-4da1-96cf-e871668385c5     1
a19afb08-c97f-46ff-b95e-bbaec4eda25b     1
a19abb76-a918-47e9-9f00-c78aba163325     1
a14fda64-ded6-4f07-beb3-1c3f6cf6c73a     1
a12121a1-3408-4cbc-ad38-1657b296e7c1     1
                                        ..
0016342a-386d-4e5c-8b84-f96ee84d3269    14
e97abc44-89c7-4771-952b-b3dab56625c3    15
ed2c2eea-8399-4779-a9ea-dafb77878965    18
3afdaa57-ecc2-4c79-94a3-d10700a299b0    23
6fde0c68-6c47-4183-b725-c5de30f90f02    25
Length: 1967, dtype: int64

In [153]:
import folium

map_osm = folium.Map(location=[50, 5],zoom_start=5)
for ind, row in m_df.iterrows():
    folium.CircleMarker(location=[row["Latitude"], row["Longitude"]],radius=2, popup=ind).add_to(map_osm)
map_osm

### 

In [154]:
# Import regions
r_df=pd.read_csv('../data/Regions.csv', sep=',', index_col='Id')
r_df.head()

Name
Id                                
0                        No Region
1               01_R01_Munich (BY)
2            01_R02_Stuttgart (BW)
3        01_R03_Frankfurt (HE, SL)
4   01_R04_Cologne (NRW_South, RP)

In [155]:
# Import container list
user_df=pd.read_csv('../data/Users_sanitized.csv', sep=',', index_col='UserId')
user_df.head()

Unnamed: 0       CreationTimestamp  \
UserId                                                                     
eecc280c-9941-41bc-b26f-d519da794a5d           0  1970-01-01 00:00:00+00   
d9d8a373-29f0-4fe5-8f17-39669083ac48           1  1970-01-01 00:00:00+00   
8f27c773-334a-4f8c-82c2-8cdec215a9bb           2  1970-01-01 00:00:00+00   
aefe87c8-f537-4bdc-954f-6833a047081b           3  1970-01-01 00:00:00+00   
6ad98c4e-18e3-4a6e-ac94-1338cb239c2d           4  1970-01-01 00:00:00+00   

                                     ClaimType  IsOfflineUser  
UserId                                                         
eecc280c-9941-41bc-b26f-d519da794a5d  consumer          False  
d9d8a373-29f0-4fe5-8f17-39669083ac48  consumer           True  
8f27c773-334a-4f8c-82c2-8cdec215a9bb  consumer           True  
aefe87c8-f537-4bdc-954f-6833a047081b  consumer           True  
6ad98c4e-18e3-4a6e-ac94-1338cb239c2d  consumer           True

In [156]:
internal_df=pd.read_csv('internal_accounts.csv', sep=',', index_col='Id')
internal_df.head()

Empty DataFrame
Columns: []
Index: [77687449-5884-4e6b-bcae-b73af6da14d0, 745bc57d-30df-4dc1-bcf7-546fbac0bd2f, 9418b689-3962-46f8-a3f9-85ab2c44109a, 1c2247cb-acaf-4b15-820a-324e8810867c, e1bc88a8-63a3-4a25-bdae-79c8bb064f9b]

In [157]:
# Import container list
# correct time zone
con_loc_df=pd.read_csv('ContainerLocations_complete.csv', sep=',', index_col='Id')
con_loc_df['OwnerSince']= pd.to_datetime(con_loc_df['OwnerSince']).dt.tz_convert(tz='Europe/Berlin')
con_loc_df['OwnerTill']= pd.to_datetime(con_loc_df['OwnerTill']).dt.tz_convert(tz='Europe/Berlin')
con_loc_df.head()

/Users/kultiginbozdemir/opt/anaconda3/envs/CO2/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3147: DtypeWarning: Columns (8,10,13,14) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
/Users/kultiginbozdemir/opt/anaconda3/envs/CO2/lib/python3.7/site-packages/numpy/lib/arraysetops.py:580: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


Unnamed: 0                               OwnerId  \
Id                                                     
23           0  2bc9794d-4076-41b0-be9d-33f2c741043f   
24           1  aeda6030-936d-454d-a398-7591b2e93338   
25           2  9d15e9cd-4112-4f84-9295-b15b2182c12b   
26           3  aeda6030-936d-454d-a398-7591b2e93338   
27           4  e5c7976f-92fb-408f-b852-01b98ba1723d   

                             NextOwnerId  \
Id                                         
23  aeda6030-936d-454d-a398-7591b2e93338   
24  9d15e9cd-4112-4f84-9295-b15b2182c12b   
25  aeda6030-936d-454d-a398-7591b2e93338   
26  d1b8228c-9972-411d-b08b-0be124e3b929   
27  c96f1b99-5db4-45ce-8294-2f31e771e35f   

                             ContainerId                       OwnerSince  \
Id                                                                          
23  e007545c-4ad9-4f05-8c28-1222c322001a 2019-09-22 21:57:42.003216+02:00   
24  e007545c-4ad9-4f05-8c28-1222c322001a 2019-09-22 21:58:10.958367+02:00   
25  e007545c-4ad9-4f05-8c28-1222c322001a 2019-09-22 21:59:32.362761+02:00   
26  e007545c-4ad9-4f05-8c28-1222c322001a 2019-09-22 21:59:43.877504+02:00   
27  963c9cb3-3c7b-4c85-be71-e808a471706b 2019-09-23 09:29:01.132709+02:00   

                          OwnerTill  IsBillableTransaction ReturnDeadline  \
Id                                                                          
23 2019-09-22 21:58:10.958117+02:00                  False            NaN   
24 2019-09-22 21:59:32.362733+02:00                   True            NaN   
25 2019-09-22 21:59:43.877483+02:00                  False            NaN   
26 2019-09-22 21:59:43.877504+02:00                  False            NaN   
27 2019-10-07 12:06:31.430389+02:00                   True            NaN   

    Status TransactionId         LastReturnReminderSent  LocationNumber  \
Id                                                                        
23       1           NaN  2020-09-03 15:02:38.794937+00               1   
24       1           NaN  2020-09-03 15:02:38.794937+00               2   
25       1           NaN  2020-09-03 15:02:38.794937+00               3   
26       1           NaN  2020-09-03 15:02:38.794937+00               4   
27       1           NaN  2020-09-03 15:02:38.794937+00               1   

   DeliveryTokenId LastStatusUpdateAt  IsSelfScan  
Id                                                 
23             NaN                NaN       False  
24             NaN                NaN       False  
25             NaN                NaN       False  
26             NaN                NaN       False  
27             NaN                NaN       False

In [158]:
# Size of df
print(con_loc_df.shape)

con_loc_df.index.is_unique



(4866672, 15)


True

In [159]:
a=con_loc_df['OwnerId'].isin(m_df.index)

unique, counts = np.unique(a, return_counts=True)
dict(zip(unique, counts))

{False: 4866672}

In [160]:
a=con_loc_df['NextOwnerId'].isin(customer_postgre_df.index)

unique, counts = np.unique(a, return_counts=True)
dict(zip(unique, counts))

{False: 564, True: 4866108}

In [161]:
a=con_loc_df['OwnerId'].isin(customer_postgre_df.index)

unique, counts = np.unique(a, return_counts=True)
dict(zip(unique, counts))

{False: 671, True: 4866001}

In [162]:
a=con_loc_df['OwnerId'].isin(internal_df.index)

unique, counts = np.unique(a, return_counts=True)
dict(zip(unique, counts))

{False: 4865158, True: 1514}

In [163]:
con_loc_df['owner_internal_']=con_loc_df['OwnerId'].isin(internal_df.index)

con_loc_df['next_owner_internal_']=con_loc_df['NextOwnerId'].isin(internal_df.index)

In [164]:
con_loc_df['owner_user_']=con_loc_df['OwnerId'].isin(customer_postgre_df.index)
con_loc_df['next_owner_user_']=con_loc_df['NextOwnerId'].isin(customer_postgre_df.index)

### ???: Both owner and nextowner are internal in the Container_Location table. How, why?

In [165]:
# 
internal_true=con_loc_df['owner_internal_']==True
next_internal_true=con_loc_df['next_owner_internal_']==True
print('Number of both owner and next_owner are internal :',con_loc_df[next_internal_true & internal_true].shape )
con_loc_df[next_internal_true & internal_true]

Number of both owner and next_owner are internal : (52, 19)


Unnamed: 0                               OwnerId  \
Id                                                         
53060        50807  d3b5f108-ab6b-4398-92e8-9c6110beab1c   
56305        53959  fd3ec806-b763-48a9-992b-09e50c8a274c   
57837        55465  fd3ec806-b763-48a9-992b-09e50c8a274c   
58905        56518  729a8000-a472-4003-931f-07566a10fa11   
58907        56520  729a8000-a472-4003-931f-07566a10fa11   
60792        58229  3e5bc859-f83e-4a00-be8c-26591fd8135c   
61023        58458  d3b5f108-ab6b-4398-92e8-9c6110beab1c   
61622        59035  ab1f32e5-f22e-46a1-b000-9fec19621bdc   
70983        67742  d3b5f108-ab6b-4398-92e8-9c6110beab1c   
80137        76630  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
94922        91168  977d42ff-8c6c-4fd8-b997-8753eaae20a0   
94930        91176  977d42ff-8c6c-4fd8-b997-8753eaae20a0   
118143      114048  1bb90ca5-00ba-46dc-9d13-bce63c30b789   
118350      114253  1bb90ca5-00ba-46dc-9d13-bce63c30b789   
137811      133305  3e5bc859-f83e-4a00-be8c-26591fd8135c   
137813      133307  3e5bc859-f83e-4a00-be8c-26591fd8135c   
179683      174104  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179685      174106  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179712      174133  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179714      174135  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179716      174137  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179718      174139  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179720      174141  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179722      174143  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179723      174144  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179724      174145  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179726      174147  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179728      174149  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179729      174150  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179734      174155  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179737      174158  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179750      174171  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179753      174174  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179767      174188  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179776      174197  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179789      174210  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179801      174222  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179836      174244  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
179840      174248  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
259961      252049  fd3ec806-b763-48a9-992b-09e50c8a274c   
260050      252138  fd3ec806-b763-48a9-992b-09e50c8a274c   
307827      298640  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
307848      298659  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
307849      298660  6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
318218      308717  fd3ec806-b763-48a9-992b-09e50c8a274c   
318220      308719  fd3ec806-b763-48a9-992b-09e50c8a274c   
327222      317446  fd3ec806-b763-48a9-992b-09e50c8a274c   
327868      318063  fd3ec806-b763-48a9-992b-09e50c8a274c   
377329      366320  1bb90ca5-00ba-46dc-9d13-bce63c30b789   
475904      461101  fd3ec806-b763-48a9-992b-09e50c8a274c   
476237      461319  1bb90ca5-00ba-46dc-9d13-bce63c30b789   
476239      461321  1bb90ca5-00ba-46dc-9d13-bce63c30b789   

                                 NextOwnerId  \
Id                                             
53060   d3b5f108-ab6b-4398-92e8-9c6110beab1c   
56305   fd3ec806-b763-48a9-992b-09e50c8a274c   
57837   fd3ec806-b763-48a9-992b-09e50c8a274c   
58905   729a8000-a472-4003-931f-07566a10fa11   
58907   729a8000-a472-4003-931f-07566a10fa11   
60792   3e5bc859-f83e-4a00-be8c-26591fd8135c   
61023   d3b5f108-ab6b-4398-92e8-9c6110beab1c   
61622   ab1f32e5-f22e-46a1-b000-9fec19621bdc   
70983   d3b5f108-ab6b-4398-92e8-9c6110beab1c   
80137   6ccdaed2-3991-4e6a-a681-d9cec70d0fa3   
94922   977d42ff-8c6c-4fd8-b997-8753eaae20a0   
94930   977d42ff-8c6c-4fd8-b997-8753eaae20a0   
118143  1bb90ca5-00ba-46dc-9d13-bce63c30b789   
118350  1bb90ca5-00ba-46dc-9d13-bce63c30b789   

### ????? : 564 Transactions in Container_Location do not have UserId foreign key. Why?

In [166]:
user_false=con_loc_df['next_user_']==False

print('Number of transactions: ',con_loc_df[user_false].shape)
con_loc_df[user_false].tail(30)

KeyError: 'next_user_'

In [ ]:
# Slice the Container Location table such that we get internal movements of containers (internal> normal user or normal user > internal)

owner_internal_true=con_loc_df['owner_internal_']==True
next_owner_internal_false=con_loc_df['next_owner_internal_']==False

internal_check_ins=con_loc_df[owner_internal_true & next_owner_internal_false] # internal check-in of containers
internal_check_ins.tail()

In [ ]:
# Slice the Container Location table such that we get internal movements of containers (normal user > internal)

owner_internal_false=con_loc_df['owner_internal_']==False
next_owner_internal_true=con_loc_df['next_owner_internal_']==True

internal_check_outs=con_loc_df[owner_internal_false & next_owner_internal_true] # internal check-out of containers
internal_check_outs.tail()

In [ ]:
con_loc_df.dtypes

In [ ]:
con_loc_df
container=con_loc_df['ContainerId']=='e007545c-4ad9-4f05-8c28-1222c322001a'
con_loc_df[container ]

In [ ]:
# import transactions
in_df=pd.read_csv('Total_checkins_all_time.csv', sep=',')


in_df['OwnerSince']= pd.to_datetime(in_df['OwnerSince'])
in_df['OwnerTill']= pd.to_datetime(in_df['OwnerTill'])

print(in_df.info())
print('shape of df ', in_df.shape)

print(min(in_df['OwnerSince']),max(in_df['OwnerSince'] ))
in_df.head(2)

In [ ]:
out_df=pd.read_csv('Total_checkouts_all_time.csv', sep=',')
out_df['OwnerSince']= pd.to_datetime(out_df['OwnerSince'])
out_df['OwnerTill']= pd.to_datetime(out_df['OwnerTill'])
print(out_df.info())
print('shape of df :', out_df.shape)
out_df.head(2)


In [ ]:
# Internal relocation movements are not recorded in check-in an check-out transactions
a=internal_df.index.isin(in_df['UserId'])

unique, counts = np.unique(a, return_counts=True)
dict(zip(unique, counts))

In [ ]:
import datetime as dt


date_time_str = '2021-05-19 17:20:15'
date_time_obj = dt.datetime.strptime(date_time_str, '%Y-%m-%d %H:%M:%S')


print ("The type of the date is now",  type(date_time_obj))
print ("The date is", date_time_obj)
print(date_time_obj.date())
date_=in_df['OwnerSince'].dt.date==date_time_obj.date()
container=in_df['ContainerId']=='2dc69c25-a17a-4fe9-8967-56e74c027468'
transaction_id=in_df[date_ & container]['TransactionId']
print(transaction_id)

In [ ]:
transaction=in_df['TransactionId']=='4222158d-02ea-4891-8961-cd5d27135376'
in_df[transaction ]

In [ ]:
#Lets follow the journey of a container...
in_df[in_df['ContainerId']=='e007545c-4ad9-4f05-8c28-1222c322001a'].sort_values(by='OwnerSince')

In [ ]:
out_df[out_df['ContainerId']=='e007545c-4ad9-4f05-8c28-1222c322001a'].sort_values(by='OwnerSince')

In [ ]:

m_df['in']=in_df.groupby(by='StoreId').size()
m_df['out']=out_df.groupby(by='StoreId').size()
m_df['delta']=in_df.groupby(by='StoreId').size()-out_df.groupby(by='StoreId').size()

In [ ]:
m_df.sort_values(by='delta', ascending=False).head()

In [ ]:
map_osm = folium.Map(location=[50, 5],zoom_start=5)
for ind, row in m_df.iterrows():
    if row['delta']<0: color='red'
    else: color='green'
    folium.CircleMarker(location=[row["Latitude"], row["Longitude"]],radius=2, popup=ind, color=color).add_to(map_osm)
map_osm

In [ ]:
m_df['delta'].hist(bins=[-5000,-2000,-1000,-500,-100,0,100,500,1000,2000,5000], figsize=(20,5))

### Cologne Analysis

In [ ]:
# Import map of German counties as geopandas file
boundaries = gpd.read_file('../data/DEU_adm/DEU_adm3.shp', encoding = "utf-8")

boundaries.plot()


In [ ]:
#Take only Cologne
cologne=boundaries[boundaries["ID_3"]==289]
cologne.plot()

In [ ]:
# create geo dataframe for Merchants

from shapely.geometry import Point

m_gdf = gpd.GeoDataFrame(m_df, geometry=gpd.points_from_xy(m_df['Longitude'], m_df['Latitude']))

m_gdf.crs = {'init' :'epsg:4326'}

m_gdf.head()

In [ ]:
#Geospatial join of merchants and Cologne polygon
vytal_cologne=gpd.sjoin( cologne, m_gdf, op='contains' ).to_crs(4326)
print(vytal_cologne.shape)
vytal_cologne.index=vytal_cologne["index_right"]
m_cologne=m_gdf[m_gdf.index.isin(vytal_cologne.index)]
m_cologne.shape

In [ ]:
m_cologne.head(2)

In [ ]:

fig, ax = plt.subplots(figsize=(12,10))
plt.title('Vytal in Cologne', fontsize=16, color='black')
cologne.plot(ax=ax)
m_cologne.plot(ax=ax, markersize=10, color='red')

In [ ]:
in_cologne=in_df[in_df["StoreId"].isin(m_cologne.index)]

out_cologne=out_df[out_df["StoreId"].isin(m_cologne.index)]

print("Size reduction:", in_df.shape, " vs.", in_cologne.shape)

print("Size reduction:", out_df.shape, " vs.", out_cologne.shape)

In [ ]:
# Gephi files
in_cologne['Source']=in_cologne['UserId']
in_cologne['Target']=in_cologne['StoreId']
in_cologne.to_csv('in_cologne.csv')

In [ ]:
out_cologne['Target']=out_cologne['UserId']
out_cologne['Source']=out_cologne['StoreId']
out_cologne.to_csv('out_cologne.csv')

### Remove Vytal accounts

In [ ]:
# Import regions
re_df=pd.read_csv('../data/Accounts Realocation of Containers .csv', sep=',', index_col='UserId')
re_df.head()

In [ ]:
in_df[in_df['UserId'].isin(re_df.index)]

In [ ]:
out_df[out_df['UserId'].isin(re_df.index)]

In [ ]:
u_df[u_df.index.isin(re_df.index)]

In [ ]:
u_df.columns

In [ ]:
out_cologne.columns

In [ ]:
out_cologne['UserId'].isin(u_df.index).value_counts()

In [ ]:
in_cologne['UserId'].isin(u_df.index).value_counts()